### DISCUSSION TITLE
ScalaFX and Scala 3

### DISCUSSION THREAD

**Discussion by MarkCLewis** [scalafx/scalafx#334] [General]
I know that there were discussions related to how to get ScalaFX to work without the DelayedInit functionality. I don't recall ever seeing how this was resolved. I'm going to begin working on the 3rd edition of my textbooks and they will use Scala 3. In the 2nd editions, I switched the GUIs and graphics from Swing to ScalaFX. Having an object extend JFXApp was great because it was simple enough for CS1. I see a discussion of using JFXApp3 with a start method. That isn't much more complex. Does that appear to be the approach that is going to be used or are other options being explored?

**Comment by jotomo** [comment]
This was just resolved, see https://github.com/scalafx/scalafx/issues/332
Demo: https://github.com/scalafx/scalafx/blob/master/scalafx-demos/src/main/scala/scalafx/JFXApp3Demo.scala#L40
Released via 15.0.1-R20 (not officially tagged yet, but already on search.maven.org)

**Reply by MarkCLewis** [reply]
Thanks. I just wanted to make sure that this was definitely the way things were going to be done going forward.

**Comment by jpsacha** [comment]
@MarkCLewis `JFXApp3` is the best/simplest suggestion I have for Scala 3 at this point. It also works with Scala 2. The idea here was to have some way of using ScalaFX with dotty/Scala 3 for those who experimenting with it. The other option is to use "standard" JavaFX `javafx.application.Application` and adding separate `Launcher` class (at least in SBT).

`JFXApp3` it is not as simple as `JFXApp` (you have to use `start()` method) and there may be potential issues when UI code is not starting in `start()` (especially that it was not needed in `JFXApp`).  I am really looking for suggestion to do it better, in a more error proof way.

Scala 3 is using new [`@main`](https://dotty.epfl.ch/docs/reference/changed-features/main-functions.html) annotation for application main method.  It is recommended over previous `App` approach.  I am not clear if we can "wrapp" `@main` to create valid JavaFX application the way `JFXApp`/`JFXApp3` are doing it. Under the hood you need to create proper `javafx.application.Application`. Scala 3 is still changing, sometime ago it was not that certain that `@main` will be part of it. I did not have enough time to look into "wrapping" `@main()` yet.

**Reply by MarkCLewis** [reply]
I'm planning to introduce the `@main` approach fairly early in CS1. Unfortunately, I know nothing about what it is doing under the hood and if it could be co-opted to set up a `javafx.application.Application`. It would be ideal if it could or if it was possible to construct something like `@mainJFX` to do that. If I find some time I could try looking into that, but as you noted, time is often hard to come by.

**Comment by jpsacha** [comment]
[ScalaFX Ensemble](https://github.com/scalafx/scalafx-ensemble) now supports cross-complication for Scala 2.13 and Scala 3 RC1. It is using regular ScalaFX release 15.0.1-R21. It is a project of a good complexity showing that using ScalaFX from Scala 3 is practical.

**Comment by jpsacha** [comment]
FYI: I was able to build ScalaFX code base with Scala 3 RC2. It did require some changes. 

I am in process of cleaning it up and and where possible changing Scala 2 code base to make it forward compatible with Scala 3. There are some places that I am still looking for better solution, for instance, `scalafx.event.EventHandlerDelegate#EventHandled` type class that will not work in Scala 3 due to new restrictions how type parameters in type classes. There are also some issues with binding constructs like: 
```scala
fill <== when (hover) choose Color.Green otherwise Color.Red
```
this should work but is some cases require declaration of intermediate variable to compile:
```scala
val helper: ObjectBinding[jfxsp.Paint] = when(hover) choose Color.Green otherwise Color.Red
fill <== helper
```
If somebody is interested in helping I can share intermediate code.

**Reply by rom1dep** [reply]
May be of interest: https://github.com/scalacenter/scala-3-migration-guide/discussions/183 :)

**Comment by jpsacha** [comment]
**Update on Scala 3 support: it is working and ready for testing**.

- Branch "Scala3" now contains code that compiles with Scala 2.12, 2.13, and 3.0.0-RC2 (commit c9c72b9137febdf8c2e26de0895fa584625dcc44). Al tests are passing. The 2.13 code was corrected to be forward compatible with Scala 3. So far there is no source code specific to Scala 3. Note, most demos were not yet ported to JFXApp3, so they will not start in Scala 3 (help needed)
- Snapshot "15.0.1-R22-SNAPSHOT" with 2.12, 2.13, and 3.0.0-RC2 binaries was published for testing
- Working example of using Scala 3 binaries, SBT setup, etc. is in the [Scala3 branch of the "scalafx-hello-world"](https://github.com/scalafx/scalafx-hello-world/commit/da2e587277623d0fc672fc4b4f27746c5e57e61d)

A rough plan for release:
* Test "15.0.1-R22-SNAPSHOT" with the Scala3 branch of [scalafx-ensemble](https://github.com/scalafx/scalafx-ensemble) 
* If scalafx-ensemble works without major issues. merge Scala3 into master branch and publish R22 with Scala 3 RC3

Let me know your thoughts

**Reply by rom1dep** [reply]
Impressive, thanks for the hard work :)

**Comment by jpsacha** [comment]
With release [16.0.0-R22](https://github.com/scalafx/scalafx/releases/tag/v.16.0.0-R22) Scala 3 is now "officially" supported (RC3 at least)
